pkg/controller/podautoscaler/horizontal.go

```diff
@@ -148,22 +148,41 @@ func (a *HorizontalController) computeReplicasForCPUUtilization(hpa *autoscaling
 		a.eventRecorder.Event(hpa, api.EventTypeWarning, "InvalidSelector", errMsg)
 		return 0, nil, time.Time{}, fmt.Errorf(errMsg)
 	}
-	currentUtilization, timestamp, err := a.metricsClient.GetCPUUtilization(hpa.Namespace, selector)
+	utilizationInfo, err := a.metricsClient.GetCPUUtilization(hpa.Namespace, selector)
 
 	// TODO: what to do on partial errors (like metrics obtained for 75% of pods).
 	if err != nil {
 		a.eventRecorder.Event(hpa, api.EventTypeWarning, "FailedGetMetrics", err.Error())
 		return 0, nil, time.Time{}, fmt.Errorf("failed to get CPU utilization: %v", err)
 	}
 
-	utilization := int32(*currentUtilization)
+	utilization := int32(utilizationInfo.Utilization)
 
-	usageRatio := float64(utilization) / float64(targetUtilization)
-	if math.Abs(1.0-usageRatio) > tolerance {
-		return int32(math.Ceil(usageRatio * float64(currentReplicas))), &utilization, timestamp, nil
+	usageRatio := utilizationInfo.Utilization / float64(targetUtilization)
+	if math.Abs(1.0-usageRatio) <= tolerance {
+		// if the change was too small, ignore it
+		return currentReplicas, &utilization, utilizationInfo.OldestTimestamp, nil
 	}
 
-	return currentReplicas, &utilization, timestamp, nil
+	// for unready pods, in the case of scale up, check to see if treating those pods as having
+	// a mock CPU utilization of zero would change things
+	if utilizationInfo.UnreadyPodsCount > 0 && usageRatio > 1.0 {
+		// check if treating unready pods as having 0 CPU usage would change things
+		newUtilization := (utilizationInfo.Utilization * float64(utilizationInfo.ReadyPodsCount)) / float64(utilizationInfo.ReadyPodsCount+utilizationInfo.UnreadyPodsCount)
+		newUsageRatio := newUtilization / float64(targetUtilization)
+
+		// simply don't scale if the new usage ratio would mean a downscale or no scale
+		if newUsageRatio < 1.0 || math.Abs(1.0-newUsageRatio) <= tolerance {
+			return currentReplicas, &utilization, utilizationInfo.OldestTimestamp, nil
+		}
+
+		// set the usage ratio to scale by to our new usage ratio (which may be a bit more conservative)
+		usageRatio = newUsageRatio
+
+		// keep the reported utilization to be whatever we retrieved from the the ready pods
+	}
+
+	return int32(math.Ceil(usageRatio * float64(currentReplicas))), &utilization, utilizationInfo.OldestTimestamp, nil
```

Comment:
Ok, so, if we split pods into running|ready ("ready") and everything else ("unready"), then I think this calculation becomes ok in the scale up case.  In the scale down case, we should always base off of running pods (otherwise, we'll kill unready/pending pods only, and the we'll have to scale down again next time).

If we consider the example from https://github.com/kubernetes/kubernetes/issues/34821#issue-183051593, and assume:

- target utilization = 100%
- current utilization = 300%
- pod 1 is ready, pods 2-10 are unready (pending or otherwise)

Then we get: `newUtilization = (300 * 1) / (1 + 9) = 300 / 10 = 30` yielding a new usage ratio of `30 / 100 = 0.3`, which gets adjusted to 1.0 (we never scale down when correcting for pending pods), leaving the desired replica count at 10.  This is slightly more conservative than simply taking the original usage ratio (3.0) and multiplying by running pods (1, giving a desired replica count of 3), but I think it's ok to be a bit more conservative when doing predictive work.  Once the pending pods become ready, the HPA may remove some if they're not doing enough work.